### PR TITLE
Reformat the data model documentation.

### PR DESCRIPTION
**Pull request by vpetrovykh**
Change the "Data Model" section to "Concepts" and update the layout.
Also describe the concepts starting from "instance" and so on.

### CODE REVIEW

**Comment by vpetrovykh** [comment]
I left the "Abstract Types" subsection mostly as is, because `anytype` is hard to allocate anywhere specific and leaving it as the only dangling bit seems odd.

File: docs/datamodel/abstract.rst [hunk 1]
```diff
@@ -4,6 +4,13 @@
 Abstract Types
 ==============
 
+These types are used in introspection as well as in definitions of
```
Comment (1st1) [inline docs/datamodel/abstract.rst:7]: Quoting https://github.com/edgedb/edgedb/discussions/2823:

> Abstract Types -> drop it. Move information about abstract scalar types to the Scalar Types section, anytuple to Collection Types, etc.

File: docs/datamodel/functions.rst [hunk 2]
```diff
@@ -29,6 +29,34 @@ cardinality:
     {2}
 
 
+User-defined Functions
+----------------------
+
+It is also possible to define custom functions. For example, consider
+a function that adds an exclamation mark ``'!'`` at the end of the
+string or produces ``'!!!'`` if the input is omitted or empty:
```
Comment (1st1) [inline docs/datamodel/functions.rst:37]: The `!!!` part is unnecessary, let's keep it stupid-simple. Just concatenating `!` is sufficient.

File: docs/datamodel/functions.rst [hunk 3]
```diff
@@ -29,6 +29,34 @@ cardinality:
...
+.. code-block:: sdl
+
+    function exclamation(word: optional str = {}) -> str
+        using ((word ++ '!') ?? '!!!');
+
+In order to make sure that the function is called when the argument is
+an empty set ``{}`` we make the argument :ref:`optional
+<ref_eql_fundamentals_optional>`. We also provide a default value of
+``{}`` if the argument is omitted entirely. Here are some results this
+function produces:
```
Comment (1st1) [inline docs/datamodel/functions.rst:48]: Yeah, the "optional" stuff is cool, but the whole purpose of this section is to say just one thing: "you can define your own functions". That's it. This section isn't about anything else, and especially it's not about `optional` arguments.

File: docs/datamodel/index.rst [hunk 4]
```diff
@@ -1,50 +1,116 @@
...
+is created with an empty database called "edgedb". If you wish to add
+more, you can use the :eql:stmt:`CREATE DATABASE` EdgeQL command.
+Conversely, the :eql:stmt:`DROP DATABASE` command removes a database.
+
+The following command will get a list of all databases present in the
+instance:
+
+.. code-block:: edgeql
+
+    SELECT sys::Database.name;
```
Comment (1st1) [inline docs/datamodel/index.rst:58]: Perhaps mention `\l` in the interactive shell?

File: docs/datamodel/index.rst [hunk 5]
```diff
@@ -1,50 +1,116 @@
...
+
+Modules
+=======
+
+Every database has a schema that fully describes it. Schema consists
+of logical units called  **modules**, which act as namespaces.
+Modules contain user-defined types, functions, etc. A module has a
+name that is unique inside a database. The same schema object name can
+be used in different modules without conflict.  For example, both
+``module1`` and ``module2`` can contain a ``User`` object type.
```
Comment (1st1) [inline docs/datamodel/index.rst:70]: Good stuff!

File: docs/datamodel/index.rst [hunk 6]
```diff
@@ -1,50 +1,116 @@
...
+
+* ``std``: standard types, functions and other declarations
+* ``schema``: types describing the :ref:`introspection <ref_eql_introspection>`
+  schema
+* ``sys``: system-wide entities, such as user roles and
+  :ref:`databases <ref_datamodel_databases>`
+* ``cfg``: configuration and settings
+* ``stdgraphql``: GraphQL-related types
+* ``math``: algebraic and statistical functions
+* ``default``: the default module for user-defined types, functions, etc.
```
Comment (1st1) [inline docs/datamodel/index.rst:85]: Add a TODO here to link to the future `Standard Library` section.

File: docs/datamodel/index.rst [hunk 7]
```diff
@@ -1,50 +1,116 @@
...
+for faster querying, and **computed** properties or links to define
+useful expressions (e.g. ``property email := .user_name ++ '@' ++
+.provider_name``).
+
+**Expression Aliases** let you use existing types under new names to
+build on them without touching the original -- both in your schema or
+on the fly inside a query.
+
+You can also create your own **functions**, strongly typed along with
+everything else in EdgeDB.
```
Comment (1st1) [inline docs/datamodel/index.rst:116]: Love this new section, great job. Now we just need a cool diagram, but we'll add it later.

**Review by 1st1 — COMMENTED** [review]
This is really great. I've left a few comments.

**Comment by 1st1** [comment]
> I left the "Abstract Types" subsection mostly as is, because anytype is hard to allocate anywhere specific and leaving it as the only dangling bit seems odd.

`anytype` should be listed in `Concepts / Functions`, perhaps add `Generic` functions for that with an example. It would be way cooler to introduce it that way, because generic functions are super cool.

File: docs/datamodel/functions.rst (hunk 2 above)
Comment (vpetrovykh) [reply docs/datamodel/functions.rst:37]: Fixed.

File: docs/datamodel/functions.rst (hunk 3 above)
Comment (vpetrovykh) [reply docs/datamodel/functions.rst:48]: FIxed.

File: docs/datamodel/index.rst (hunk 4 above)
Comment (vpetrovykh) [reply docs/datamodel/index.rst:58]: Added.

File: docs/datamodel/index.rst (hunk 6 above)
Comment (vpetrovykh) [reply docs/datamodel/index.rst:89]: Added

**Review by 1st1 — APPROVED** [review]
OK to go and let's continue refactoring the docs

**Pull request merged**